plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/FileSystemExchange.java

```diff
@@ -199,53 +208,37 @@ private List<ExchangeSourceHandle> createExchangeSourceHandles()
         return result.build();
     }
 
-    private URI getCommittedAttemptPath(int taskPartitionId)
+    private ListenableFuture<Multimap<Integer, FileStatus>> getCommittedPartitions(int taskPartitionId)
     {
-        URI sinkOutputBasePath = getTaskOutputDirectory(taskPartitionId);
-        try {
-            List<URI> attemptPaths = exchangeStorage.listDirectories(sinkOutputBasePath);
-            checkState(!attemptPaths.isEmpty(), "No attempts found under sink output path %s", sinkOutputBasePath);
-
-            return attemptPaths.stream()
-                    .filter(this::isCommitted)
-                    .findFirst()
-                    .orElseThrow(() -> new IllegalStateException(format("No committed attempts found under sink output path %s", sinkOutputBasePath)));
-        }
-        catch (IOException e) {
-            throw new UncheckedIOException(e);
-        }
-    }
-
-    private boolean isCommitted(URI attemptPath)
-    {
-        URI commitMarkerFilePath = attemptPath.resolve(COMMITTED_MARKER_FILE_NAME);
-        try {
-            return exchangeStorage.exists(commitMarkerFilePath);
-        }
-        catch (IOException e) {
-            throw new UncheckedIOException(e);
-        }
-    }
-
-    private Multimap<Integer, FileStatus> getCommittedPartitions(URI committedAttemptPath)
-    {
-        try {
-            List<FileStatus> partitionFiles = exchangeStorage.listFiles(committedAttemptPath)
-                    .stream()
-                    .filter(file -> file.getFilePath().endsWith(DATA_FILE_SUFFIX))
-                    .collect(toImmutableList());
-            ImmutableMultimap.Builder<Integer, FileStatus> result = ImmutableMultimap.builder();
-            for (FileStatus partitionFile : partitionFiles) {
-                Matcher matcher = PARTITION_FILE_NAME_PATTERN.matcher(new File(partitionFile.getFilePath()).getName());
-                checkState(matcher.matches(), "Unexpected partition file: %s", partitionFile);
-                int partitionId = Integer.parseInt(matcher.group(1));
-                result.put(partitionId, partitionFile);
-            }
-            return result.build();
-        }
-        catch (IOException e) {
-            throw new UncheckedIOException(e);
-        }
+        URI sinkOutputPath = getTaskOutputDirectory(taskPartitionId);
+        return Futures.transform(
+                exchangeStorage.listFilesRecursively(sinkOutputPath),
```

Comment:
nit: would be nice to have a comment which explains how directory structure looks like, so it is more obvious to the reader what is happening here.